extensions/amp-next-page/0.1/amp-next-page.js

```diff
@@ -286,107 +86,24 @@ export class AmpNextPage extends AMP.BaseElement {
-      }
-
-      const prerenderHeight = PRERENDER_VIEWPORT_COUNT * viewportSize.height;
-      if (elementBox.bottom - viewportBox.bottom < prerenderHeight) {
-        this.documentQueued_ = true;
-        this.appendNextArticle_();
-      }
-    });
-  }
-
-  /**
-   * Handles updates from the {@code PositionObserver} indicating a change in
-   * the position of a recommendation unit in the viewport.
-   * @param {number} i Index of the documentRef this recommendation unit is
-   *     attached to.
-   * @param
-   * {!../../../src/service/position-observer/position-observer-worker.PositionInViewportEntryDef}
-   *     position Position of the current recommendation unit in the viewport.
-   */
-  positionUpdate_(i, position) {
-    // We're only interested when the recommendations exit the viewport
-    if (position.positionRect !== null) {
-      return;
+      separator = this.win.document.createElement('div');
     }
 
-    if (position.relativePos === 'top') {
-      const documentRef = this.documentRefs_[i + 1];
-      this.triggerAnalyticsEvent_('amp-next-page-scroll',
-          documentRef.ampUrl, this.activeDocumentRef_.ampUrl);
-      this.setActiveDocument_(documentRef);
-    } else if (position.relativePos === 'bottom') {
-      const documentRef = this.documentRefs_[i];
-      this.triggerAnalyticsEvent_('amp-next-page-scroll-back',
-          documentRef.ampUrl, this.activeDocumentRef_.ampUrl);
-      this.setActiveDocument_(documentRef);
-    }
+    this.service_.register(this.element, config, separator);
+    this.service_.setAppendPageHandler(element => this.appendPage_(element));
   }
 
-  /**
-   * Sets the specified document as active, updating the document title and URL.
-   * @param {!DocumentRef} documentRef Reference to the document to set as
-   *     active.
-   */
-  setActiveDocument_(documentRef) {
-    const amp = documentRef.amp;
-    this.win.document.title = amp.title || '';
-    if (this.win.history.replaceState) {
-      const url = parseUrl(documentRef.ampUrl);
-      this.win.history.replaceState({}, amp.title, url.pathname);
-    }
-
-    this.activeDocumentRef_ = documentRef;
-
-    // TODO(peterjosling): Send request to viewer with title/URL
-    // TODO(emarchiori): Consider updating position fixed elements.
-  }
-
-  /**
-   * Wrapper around {@code triggerAnalyticsEvent}.
-   * @param {string} eventType
-   * @param {string} toURL The new url after the event.
-   * @param {string=} fromURL The old URL before the event.
-   */
-  triggerAnalyticsEvent_(eventType, toURL, fromURL) {
-    fromURL = fromURL || '';
-
-    const vars = {toURL, fromURL};
-    triggerAnalyticsEvent(this.element, eventType, vars);
+  appendPage_(element) {
+    return this.mutateElement(() => this.element.appendChild(element));
   }
 }
 
-AMP.registerElement(TAG, AmpNextPage, CSS);
+AMP.extension(TAG, '0.1', AMP => {
+  const service = new NextPageService();
+  AMP.registerServiceForDoc(SERVICE_ID, () => service);
```

Comment:
No, because then it ends up creating multiple instances of the service if any of the shadow documents use the extension. Registering the service against the window instead of the document would work, but the AMP global only has `registerServiceForDoc`, and not `registerServiceBuilder`.

Reply:
There are some extensions that use `registerServiceBuilder` directly without referencing it from the AMP document. But we can investigate that later and ship this PR now if you're ready for us to merge it.